File: rewrite-prototype/src/game/economy.ts
import type { UpgradeCatalog, UpgradeSlot } from "./types";

export function triangular(step: number): number {
  return (step * (step + 1)) / 2;
}

export function nextWeaponUpgradeCost(basePrice: number, currentPower: number): number {
  return basePrice * triangular(currentPower);
}

export function maxWeaponPower(): number {
  return 5;
}

export function slotLabel(slot: UpgradeSlot): string {
  switch (slot) {
    case "front":
      return "Front Weapon";
    case "rear":
      return "Rear Weapon";
    case "shield":
      return "Shield";
    case "generator":
      return "Generator";
    case "leftSidekick":
      return "Left Sidekick";
    case "rightSidekick":
      return "Right Sidekick";
  }
}

export function getCatalogItemBasePrice(
  catalog: UpgradeCatalog,
  slot: UpgradeSlot,
  itemId: string,
): number {
  switch (slot) {
    case "front":
      return catalog.frontWeapons.find((item) => item.id === itemId)?.basePrice ?? 0;
    case "rear":
      return catalog.rearWeapons.find((item) => item.id === itemId)?.basePrice ?? 0;
    case "shield":
      return catalog.shields.find((item) => item.id === itemId)?.basePrice ?? 0;
    case "generator":
      return catalog.generators.find((item) => item.id === itemId)?.basePrice ?? 0;
    case "leftSidekick":
    case "rightSidekick":
      return catalog.sidekicks.find((item) => item.id === itemId)?.basePrice ?? 0;
  }
}

File: TyrianClone/TyrianClone/GameCore/Economy/Economy.swift
import Foundation

enum Economy {
    static func triangular(_ step: Int) -> Int {
        (step * (step + 1)) / 2
    }

    static func nextWeaponUpgradeCost(basePrice: Int, currentPower: Int) -> Int {
        basePrice * triangular(currentPower)
    }

    static func maxWeaponPower() -> Int {
        11
    }

    static func slotLabel(_ slot: UpgradeSlot) -> String {
        switch slot {
        case .ship:
            "Ship"
        case .front:
            "Front Weapon"
        case .rear:
            "Rear Weapon"
        case .shield:
            "Shield"
        case .generator:
            "Generator"
        case .leftSidekick:
            "Left Sidekick"
        case .rightSidekick:
            "Right Sidekick"
        }
    }

    static func basePrice(in catalog: UpgradeCatalog, slot: UpgradeSlot, itemID: String) -> Int {
        switch slot {
        case .ship:
            catalog.ships.first(where: { $0.id == itemID })?.shopCost ?? 0
        case .front:
            catalog.frontWeapons.first(where: { $0.id == itemID })?.basePrice ?? 0
        case .rear:
            catalog.rearWeapons.first(where: { $0.id == itemID })?.basePrice ?? 0
        case .shield:
            catalog.shields.first(where: { $0.id == itemID })?.basePrice ?? 0
        case .generator:
            catalog.generators.first(where: { $0.id == itemID })?.basePrice ?? 0
        case .leftSidekick, .rightSidekick:
            catalog.sidekicks.first(where: { $0.id == itemID })?.basePrice ?? 0
        }
    }
}
